src/libopensc/card-myeid.c

```diff
@@ -1185,6 +1261,149 @@ static int myeid_decipher(struct sc_card *card, const u8 * crgram,
 }
 
 
+static int myeid_wrap_key(struct sc_card *card, u8 *out, size_t outlen)
+{
+	struct sc_context *ctx;
+	struct sc_apdu apdu;
+	u8 rbuf[SC_MAX_APDU_BUFFER_SIZE];
+	int r;
+
+	assert(card != NULL);
+	ctx = card->ctx;
+	LOG_FUNC_CALLED(ctx);
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	   P1:  0x84  Resp: Return a cryptogram
+	 * P2:  0x00  The data field is absent */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0x2A, 0x84, 0x00);
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = 0;
+	apdu.lc = 0;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, r, "APDU transmit failed");
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(ctx, r, "wrap key failed");
+
+	if (apdu.resplen <= outlen && out != NULL)
+		memcpy(out, apdu.resp, apdu.resplen);
```

Comment:
Should we return an error if the buffer is too small?

Reply:
I copy and modify my answer from an earlier conversation:

The idea has been, that in case where all goes well but only the buffer is too small, to return the required buffer length to caller, so that caller can allocate a long enough buffer. This convention seems to be used at least in decipher. I looked at other drivers too and they work like this. On the other hand, compute_signature returns SC_ERROR_BUFFER_TOO_SMALL if the buffer cannot hold the data returned from card. Do you know why there are different conventions in these two functions?

PKCS#11 defines returning buffer length in section 11.2: Conventions for functions returning output in a variable-length buffer
How does it work with OpenSC if you call C_Sign with pSignature = NULL?

In the wrapkey case, insufficient buffer is handled in upper level and accurate required buffer size is returned to PKCS#11 level according to the PKCS#11 convention. If the return buffer is not NULL, caller finally gets CKR_BUFFER_TOO_SMALL.

Reply:
OpenSC should be haved as defined by PKCS#11:

>The signing operation must have been initialized with C_SignInit. A call to C_Sign always terminates the active signing operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the signature.

The standard doesn't give such details about `C_WrapKey`, so I think your implementation should be fine